projects/core/src/anonymous-consents/http-interceptors/anonymous-consents-interceptor.ts

```diff
@@ -0,0 +1,84 @@
+import {
+  HttpEvent,
+  HttpHandler,
+  HttpInterceptor,
+  HttpRequest,
+  HttpResponse,
+} from '@angular/common/http';
+import { Injectable } from '@angular/core';
+import { Observable } from 'rxjs';
+import { switchMap, take, tap, withLatestFrom } from 'rxjs/operators';
+import { AuthService } from '../../auth/index';
+import { AnonymousConsent } from '../../model/index';
+import { AnonymousConsentsService } from '../facade/anonymous-consents.service';
+
+export const ANONYMOUS_CONSENTS_HEADER = 'X-Anonymous-Consents';
+
+@Injectable()
+export class AnonymousConsentsInterceptor implements HttpInterceptor {
+  constructor(
+    private anonymousConsentsService: AnonymousConsentsService,
+    private authService: AuthService
+  ) {}
+
+  intercept(
+    request: HttpRequest<any>,
+    next: HttpHandler
+  ): Observable<HttpEvent<any>> {
+    return this.anonymousConsentsService.getAnonymousConsents().pipe(
+      take(1),
+      withLatestFrom(this.authService.isUserLoggedIn()),
+      switchMap(([consents, isUserLoggedIn]) => {
+        const clonedRequest = this.handleRequest(consents, request);
+        return next.handle(clonedRequest).pipe(
+          tap(event => {
+            if (event instanceof HttpResponse) {
+              this.handleResponse(
+                event.headers.get(ANONYMOUS_CONSENTS_HEADER),
+                isUserLoggedIn
+              );
+            }
+          })
+        );
+      })
+    );
+  }
+
+  private handleResponse(rawConsents: string, isUserLoggedIn: boolean): void {
+    if (rawConsents && !isUserLoggedIn) {
+      const consents = this.decodeAndDeserialize(rawConsents);
+      this.anonymousConsentsService.setAnonymousConsents(consents);
+    }
+  }
+
+  private decodeAndDeserialize(rawConsents: string): AnonymousConsent[] {
+    const decoded = decodeURIComponent(rawConsents);
+    const unserialized = JSON.parse(decoded) as AnonymousConsent[];
+    return unserialized;
+  }
+
+  private handleRequest(
+    consents: AnonymousConsent[],
+    request: HttpRequest<any>
+  ): HttpRequest<any> {
+    if (!consents) {
+      return request;
+    }
+
+    const rawConsents = this.serializeAndEncode(consents);
+    const headers = request.headers.append(
+      ANONYMOUS_CONSENTS_HEADER,
+      rawConsents
+    );
+    return request.clone({ headers });
```

Comment:
don't know the performance or which is better, but why wouldn't you use the 
setHeaders: { 
ANONYMOUS_CONSENTS_HEADER: rawConsents
}

Reply:
Interceptors should be built in an immutable way, so the request has to be cloned.